sbi/inference/trainers/base.py

```diff
@@ -306,6 +319,198 @@ def train(
+        sample_with: Literal[
+            "mcmc", "rejection", "vi", "importance", "direct", "sde", "ode"
+        ],
+        **kwargs,
+    ) -> NeuralPosterior:
+        r"""Method for building posteriors.
+
+        This method serves as a base method for constructing a posterior based
+        on a given estimator and prior. The posterior can be sampled using one of
+        several inference methods specified by `sample_with`.
+        Args:
+            estimator: The estimator that the posterior is based on.
+            prior: A probability distribution that expresses prior knowledge about the
+                parameters, e.g. which ranges are meaningful for them. Must be a PyTorch
+                distribution, see FAQ for details on how to use custom distributions.
+            sample_with: The inference method to use. Must be one of:
+                - "mcmc"
+                - "rejection"
+                - "vi"
+                - "importance"
+                - "direct"
+                - "sde"
+                - "ode"
+            **kwargs: Additional method-specific parameters.
+        Returns:
+            NeuralPosterior object.
+        """
+        if estimator is not None and not isinstance(
+            estimator, (ConditionalEstimator, RatioEstimator)
+        ):
+            raise TypeError(
+                "estimator must be ConditionalEstimator or RatioEstimator,",
+                " got {type(estimator).__name__}",
+            )
+
+        if prior is None:
+            cls_name = self.__class__.__name__
+            assert (
+                self._prior is not None
+            ), f"""You did not pass a prior. You have to pass the prior either at
+                initialization `inference = {cls_name}(prior)` or to `.build_posterior
+                (prior=prior)`."""
+            prior = self._prior
+        else:
+            check_prior(prior)
+
+        if estimator is None:
+            estimator = self._neural_net
+            # If internal net is used device is defined.
+            device = self._device
+        else:
+            # Otherwise, infer it from the device of the net parameters.
+            device = str(next(estimator.parameters()).device)
+
+        assert estimator is not None, "Expected a valid estimator object, but got None."
+
+        potential_fn, theta_transform = self._get_potential_function(prior, estimator)
+        self._posterior = self._create_posterior(
+            potential_fn,
+            theta_transform,
+            sample_with,
+            prior,
+            device,
+            estimator,
+            **kwargs,
+        )
+
+        # Store models at end of each round.
+        self._model_bank.append(deepcopy(self._posterior))
+
+        return deepcopy(self._posterior)
+
+    @abstractmethod
+    def _get_potential_function(
+        self,
+        prior: Distribution,
+        estimator: Union[RatioEstimator, ConditionalEstimator],
+    ) -> Tuple:
```

Comment:
the return type should be more specific, e.g., `Tuple[BasePotential, TorchTransform]`?